src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/ControlProblemElements.h

```diff
@@ -0,0 +1,164 @@
+/**
+ * @file ControlProblemElements.h
+ * @authors Giulio Romualdi
+ * @copyright 2019 Istituto Italiano di Tecnologia (IIT). This software may be modified and
+ * distributed under the terms of the GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#ifndef BIPEDAL_LOCOMOTION_CONTROLLERS_OPTIMAL_CONTROL_UTILITIES_CONTROL_PROBLEM_ELEMENTS_H
+#define BIPEDAL_LOCOMOTION_CONTROLLERS_OPTIMAL_CONTROL_UTILITIES_CONTROL_PROBLEM_ELEMENTS_H
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <iDynTree/Model/Indices.h>
+#include <iDynTree/Core/Utils.h>
+
+#include <iDynTree/Core/MatrixDynSize.h>
+
+#include <iDynTree/Core/Rotation.h>
+#include <iDynTree/Core/VectorDynSize.h>
+#include <iDynTree/KinDynComputations.h>
+
+namespace BipedalLocomotionControllers
+{
+
+namespace OptimalControlUtilities
+{
+
+/**
+ * The struct frame defines a Frame in a ControlProblemElement
+ */
+struct Frame
+{
+    iDynTree::FrameIndex indexInModel; /**< Index of the frame in the Model */
+
+    /** Index and size of the element in the ControlProblemElement */
+    iDynTree::IndexRange indexRangeInElement{iDynTree::IndexRange::InvalidRange()};
+};
+
+/**
+ * FrameNames wraps a std::pair<std::string, std::string>. It contains the label
+ * associated to the frame in the VariableHandler end the name of the frame in the model
+ */
+class FrameNames : public std::pair<std::string, std::string>
+{
+    // obscure first and second attribute
+    using std::pair<std::string, std::string>::first;
+    using std::pair<std::string, std::string>::second;
+
+public:
+
+    // use the std::pair constructors
+    using std::pair<std::string, std::string>::pair;
+
+    std::string& label();
+    const std::string& label() const;
+    std::string& nameInModel();
+    const std::string& nameInModel() const;
+};
+
+/**
+ * ControlProblemElement describes a general control problem element. The
+ * element is considered linear w.r.t the unknown variable.
+ * i.e. \f$ A x \f$ where \f$ A \f$ is the element matrix and \f$ x \f$ the
+ * unknown variable
+ */
+class ControlProblemElement
+{
+protected:
+    std::shared_ptr<iDynTree::KinDynComputations> m_kinDynPtr; /**< KinDyn pointer object */
+    iDynTree::MatrixDynSize m_A; /**< Element Matrix */
+    std::string m_name; /**< Name of the element */
+
+public:
+    /**
+     * Constructor
+     * @param kinDyn KinDynComputations object
+     */
+    ControlProblemElement(std::shared_ptr<iDynTree::KinDynComputations> kinDyn);
```

Comment:
Since this class has no "setters" I would suggest putting the constructor as protected to avoid the user to instantiate it. I would do it also for the ``ControlTask`` and ``InequalityConstraintElement``.

Reply:
* InequalityConstraintElement: e4e777fbc5fa1a6289e9bc2add0ab47aeb829b42
* ControlTask 973ebf5c57928fb78a13e505a233ea30aa0a1582
* ControlProblemElement a0a3ef4775b6d2ebfd589c270d3ce23af714b065